File: cpp/strategy/DuckSimulator/behavior/QuackBehavior.cpp
#include "iostream"
#include <string>
using namespace std;

class QuackBehavior{

public:
    virtual void quack() = 0;
};

class Quack : public QuackBehavior {

public:
    Quack(string duckName) : duckName(duckName) {};
    void quack() override {
        cout << duckName << " is quacking" << endl;
    };

private:
    string duckName;
};

class Squeak : public QuackBehavior {

public:
    Squeak(string duckName) : duckName(duckName) {};
    void quack() override { cout << duckName << " is squeaking" << endl; };

private:
    string duckName;
};

class MuteQuack : public QuackBehavior {

public:
    MuteQuack(string duckName) : duckName(duckName) {};
    void quack() override { cout << duckName << " is not quacking" << endl; };

private:
    string duckName;
};

File: javascript/strategy/DuckSimulator/behavior/QuackBehavior.js

export class QuackBehavior {
    quack() {}
}

export class Quack extends QuackBehavior {

    // private vars
    #duck_name

    constructor(duck_name) {
        super();
        this.#duck_name = duck_name
    }

    quack() {
        super.quack();
        console.log(this.#duck_name + " is quacking")
    }
}

export class Squeak extends QuackBehavior {

    // private vars
    #duck_name

    constructor(duck_name) {
        super();
        this.#duck_name = duck_name
    }

    quack() {
        super.quack();
        console.log(this.#duck_name + " is squeaking")
    }
}

export class MuteQuack extends QuackBehavior {

    // private vars
    #duck_name

    constructor(duck_name) {
        super();
        this.#duck_name = duck_name
    }

    quack() {
        super.quack();
        console.log(this.#duck_name + " is not quacking")
    }
}